README.md
## syslog-message - Syslog message parser

%%VERSION%%

This is a library for parsing and generating RFC 3164 compatible Syslog messages.

## Documentation

[![Build Status](https://img.shields.io/endpoint?url=https%3A%2F%2Fci.ocamllabs.io%2Fbadge%2Fverbosemode%2Fsyslog-message%2Fmain&logo=ocaml)](https://ci.ocamllabs.io/github/verbosemode/syslog-message)

[API documentation](https://verbosemode.github.io/syslog-message/doc/) is available online.

```ocaml
match Ptime.of_date_time ((1970, 1, 1), ((0, 0, 0), 0)) with
| Some ts -> Syslog_message.decode ~ctx:{timestamp=ts; hostname="-"; set_hostname=false} "<133>Oct  3 15:51:21 server001: foobar"
| None -> failwith "Failed to parse Syslog message";;
- : Syslog_message.t option =
Some {Syslog_message.facility = Syslog_message.Local0; severity = Syslog_message.Notice; timestamp = <abstr>;
  hostname = "server001"; message = "foobar"}
```

## Installation

This library can be installed with `opam`: `opam install syslog-message`

## Testing

A test suite using qcheck is provided: `opam install --build-test syslog-message`


CHANGES.md
## 1.2.0 (2023-06-03)

* remove dependency on astring (#28, @hannesm)
* use OCaml-CI (#27, @Leonidas-from-XIV, @verbosemode)
* bump required dune version to 2.0
* raise lower bound to ocaml 4.08.0; remove rresult dependency (#25, by @hannesm)

## 1.1.0 (2019-04-14)

* additonal conversion and pretty printer functions (#23, by @vbmithr)

## 1.0.0 (2018-10-14)

* Warning: encode function no longer truncates messages to 1024 bytes by default
* split message part into tag and content (#20, by @hannesm)
* use result types instead of option (#20, by @hannesm)
* remove transport-dependent length check from encode (#20, by @hannesm)
* switch build system to Dune (#19, by @dra27)
* add encode_local for sending to local syslog (#17, by @dra27)
* forgot to thank @hannesm, @Leonidas-from-XIV for past contributions

## 0.0.2 (2016-10-29)

* simplify API: no set_hostname, hostname anymore #11
* introduce Rfc3164_timestamp module #11
* parse is now decode, to_string encode #11
* pp_string is now to_string #11
* provide pp : Format.formatter -> t -> unit
* remove int_to_severity/severity_to_int/int_to_facility/facility_to_int #11
* use topkg instead of oasis
* cleanups #8 #9

## 0.0.1 (2016-03-24)

* Initial release supporting RFC 3164


test/tests.ml
open Syslog_message.Rfc_5424

(* [mirageOsPowerSupply@49836 status="ok" name="Camelus Dromedarius"] *)
let test1 () =
  let sd_id1 = Sd_id.create "mirageOsPowerSupply" "49836" in
  let sd_elt1 = Sd_element.create sd_id1 in

  let sd_param = Sd_param.create "name" (`Utf8 "Camelus Dromedarius") in
  let sd_elt1 = Sd_element.add sd_param sd_elt1 in
  let sd_param = Sd_param.create "status" (`Utf8 "ok") in
  let sd_elt1 = Sd_element.add sd_param sd_elt1 in

  let sd = Structured_data.add sd_elt1 (Structured_data.empty) in

  (* let syslogmsg = { *)
  (*   facility=Syslog_message.Local0; *)
  (*   severity=Syslog_message.Notice; *)
  (*   version=1; *)
  (*   timestamp= *)
  (*     (match Ptime.of_date_time ((2017, 1, 1), ((1, 1, 1), 0)) with *)
  (*     | None -> assert false *)
  (*     | Some t -> Some t *)
  (*     ); *)
  (*   hostname=Some "alepale"; *)
  (*   app_name=None; *)
  (*   procid=Some "42"; *)
  (*   msgid=None; *)
  (*   structured_data=Some sd; *)
  (*   msg=`Ascii "test message"; *)
  (* } *)
  (* in *)
  (* encode syslogmsg *)

  let timestamp = Ptime.of_date_time ((2017, 1, 1), ((1, 1, 1), 0)) in
  let open Syslog_message in
    create ~facility:Local0 ~severity:Notice
      ~timestamp ~hostname:(Some "alepale") ~procid:(Some "42")
      ~structured_data:(Some sd) (`Ascii "test message")

let test1_result =
  "<133>1 2017-01-01T01:01:01Z alepale - 42 - [mirageOsPowerSupply@49836 \
  status=\"ok\" name=\"Camelus Dromedarius\"] test message"

let test_message () =
  Alcotest.(check string) "Message encoding" test1_result (test1 ())

let test_set = [
  "Example message", `Quick, test_message;
]

let () =
  Alcotest.run "Running even more tests" [
    "test_set", test_set;
  ]


src/test_syslog_message.ml
let () =
  let suite = [] in
  ignore @@ QCheck.run_tests suite


test/test_syslog_message.ml
open Syslog_message

module QCheck_legacy = struct
(* I've copied this code over from QCheck 0.4 to preserve the way, strings are
 * generated.
 *
 * Copyright (c) 2013, Simon Cruanes
 * All rights reserved.
 *
 * License (BSD): https://github.com/c-cube/qcheck/blob/6e002d5b3afbb32e364b5fa4b7e3f6e32b8d0dde/LICENSE
 * Source: https://github.com/c-cube/qcheck/blob/6e002d5b3afbb32e364b5fa4b7e3f6e32b8d0dde/qCheck.ml *)
  let alpha st =
    Char.chr (Char.code 'a' + Random.State.int st (Char.code 'z' - Char.code 'a'))

  let string_len len st =
    let n = len st in
    assert (n>=0);
    let b = Buffer.create n in
    for _i = 0 to n-1 do
      Buffer.add_char b (alpha st)
    done;
    Buffer.contents b

  let string_g =
    QCheck.Gen.string_size ~gen:alpha (QCheck.Gen.int_range 0 10)

  let string =
    QCheck.(string_gen_of_size (QCheck.Gen.int_range 0 10) alpha)
end

(* 8 severities * 23 facilities *)
let priority = QCheck.int_bound 184
let priority_g = QCheck.Gen.int_bound 184

let ptime_g =
  let open QCheck.Gen in
  int_bound (int_of_float @@ 2. ** 29.)
  >|= fun n ->
    n
    |> Ptime.Span.of_int_s
    |> Ptime.of_span

let pp_ptime = function
  | None -> "<None>"
  | Some pt -> Rfc3164_Timestamp.encode pt

let ptime = QCheck.make ~print:pp_ptime ptime_g

let valid_data_succeeds =
  let open QCheck in
  Test.make ~count:100
  ~name:"generating valid data gets a reasonable result"
  (triple priority ptime QCheck_legacy.string)
  @@ fun (pri, pt, host) ->
    assume (pt <> None && String.length host > 1);
    match pt with
    | None -> false
    | Some pt ->
      let ctx = { timestamp = pt; hostname = ""; set_hostname = false } in
      let msg =
        Printf.sprintf "<%d>%s %s: Whatever" pri (Rfc3164_Timestamp.encode pt) host
      in
      match decode ~ctx msg with
      | Error _ -> false
      | Ok parsed ->
        let ((_, m, d), ((hh, mm, ss), _)) = Ptime.to_date_time parsed.timestamp in
        let ((_, m', d'), ((hh', mm', ss'), _)) = Ptime.to_date_time pt in
        m = m' && d = d' && hh = hh' && mm = mm' && ss = ss' &&
        parsed.hostname = host

let invalid_timestamp =
  let open QCheck in
  let pp = Print.(quad int pp_ptime string string) in
  Test.make ~count:100
  ~name:"parser substitutes the timestamp when it can't be parsed"
  (make ~print:pp Gen.(quad priority_g ptime_g QCheck_legacy.string_g
    QCheck_legacy.string_g))
  @@ fun (pri, valid, invalid, host) ->
    assume (valid <> None);
    match valid with
    | None -> false
    | Some valid ->
      let msg =
        Printf.sprintf "<%d>%s %s: Whatever" pri invalid host
      in
      let ctx = { timestamp = valid; hostname = ""; set_hostname = false } in
      match decode ~ctx msg with
      | Error _ -> false
      | Ok parsed -> parsed.timestamp = valid

let invalid_data_fails =
  QCheck.Test.make ~count:100
  ~name:"putting in invalid data always fails"
  QCheck.string
  @@ fun msg ->
    let ctx = { timestamp = Ptime.epoch; hostname = ""; set_hostname = false } in
    match decode ~ctx msg with
    | Error _ -> true
    | Ok _ -> false

let () =
  let suite = [invalid_data_fails; valid_data_succeeds; invalid_timestamp] in
  QCheck_runner.run_tests_main suite


src/syslog_message.ml
let ( let* ) = Result.bind

type facility =
  | Kernel_Message
  | User_Level_Messages
  | Mail_System
  | System_Daemons
  | Security_Authorization_Messages
  | Messages_Generated_Internally_By_Syslogd
  | Line_Printer_Subsystem
  | Network_News_Subsystem
  | UUCP_subsystem
  | Clock_Daemon
  | Security_Authorization_Messages_10
  | Ftp_Daemon
  | Ntp_Subsystem
  | Log_Audit
  | Log_Alert
  | Clock_Daemon_15
  | Local0
  | Local1
  | Local2
  | Local3
  | Local4
  | Local5
  | Local6
  | Local7

let int_of_facility = function
  | Kernel_Message -> 0
  | User_Level_Messages -> 1
  | Mail_System -> 2
  | System_Daemons -> 3
  | Security_Authorization_Messages -> 4
  | Messages_Generated_Internally_By_Syslogd -> 5
  | Line_Printer_Subsystem -> 6
  | Network_News_Subsystem -> 7
  | UUCP_subsystem -> 8
  | Clock_Daemon -> 9
  | Security_Authorization_Messages_10 -> 10
  | Ftp_Daemon -> 11
  | Ntp_Subsystem -> 12
  | Log_Audit -> 13
  | Log_Alert -> 14
  | Clock_Daemon_15 -> 15
  | Local0 -> 16
  | Local1 -> 17
  | Local2 -> 18
  | Local3 -> 19
  | Local4 -> 20
  | Local5 -> 21
  | Local6 -> 22
  | Local7 -> 23

let facility_of_int = function
  | 0  -> Some Kernel_Message
  | 1  -> Some User_Level_Messages
  | 2  -> Some Mail_System
  | 3  -> Some System_Daemons
  | 4  -> Some Security_Authorization_Messages
  | 5  -> Some Messages_Generated_Internally_By_Syslogd
  | 6  -> Some Line_Printer_Subsystem
  | 7  -> Some Network_News_Subsystem
  | 8  -> Some UUCP_subsystem
  | 9  -> Some Clock_Daemon
  | 10 -> Some Security_Authorization_Messages_10
  | 11 -> Some Ftp_Daemon
  | 12 -> Some Ntp_Subsystem
  | 13 -> Some Log_Audit
  | 14 -> Some Log_Alert
  | 15 -> Some Clock_Daemon_15
  | 16 -> Some Local0
  | 17 -> Some Local1
  | 18 -> Some Local2
  | 19 -> Some Local3
  | 20 -> Some Local4
  | 21 -> Some Local5
  | 22 -> Some Local6
  | 23 -> Some Local7
  | _  -> None

let string_of_facility = function
  | Kernel_Message -> "kern"
  | User_Level_Messages -> "user"
  | Mail_System -> "mail"
  | System_Daemons -> "daemon"
  | Security_Authorization_Messages -> "security/auth"
  | Messages_Generated_Internally_By_Syslogd -> "syslog"
  | Line_Printer_Subsystem -> "lpr"
  | Network_News_Subsystem -> "news"
  | UUCP_subsystem -> "uucp"
  | Clock_Daemon -> "clock"
  | Security_Authorization_Messages_10 -> "security/auth-10"
  | Ftp_Daemon -> "ftp"
  | Ntp_Subsystem -> "ntp"
  | Log_Audit -> "log-audit"
  | Log_Alert -> "log-alert"
  | Clock_Daemon_15 -> "clock-15"
  | Local0 -> "local0"
  | Local1 -> "local1"
  | Local2 -> "local2"
  | Local3 -> "local3"
  | Local4 -> "local4"
  | Local5 -> "local5"
  | Local6 -> "local6"
  | Local7 -> "local7"

let pp_print_facility ppf f =
  Format.fprintf ppf "%s" (string_of_facility f)

type severity =
  | Emergency
  | Alert
  | Critical
  | Error
  | Warning
  | Notice
  | Informational
  | Debug

let int_of_severity = function
  | Emergency -> 0
  | Alert -> 1
  | Critical -> 2
  | Error -> 3
  | Warning -> 4
  | Notice -> 5
  | Informational -> 6
  | Debug -> 7

let severity_of_int = function
  | 0 -> Some Emergency
  | 1 -> Some Alert
  | 2 -> Some Critical
  | 3 -> Some Error
  | 4 -> Some Warning
  | 5 -> Some Notice
  | 6 -> Some Informational
  | 7 -> Some Debug
  | _ -> None

let string_of_severity = function
  | Emergency -> "emerg"
  | Alert -> "alert"
  | Critical -> "crit"
  | Error -> "err"
  | Warning -> "warning"
  | Notice -> "notice"
  | Informational -> "info"
  | Debug -> "debug"

let pp_print_severity ppf s =
  Format.fprintf ppf "%s" (string_of_severity s)

type ctx = {
  timestamp    : Ptime.t;
  hostname     : string;
  set_hostname : bool
}

type t = {
  facility  : facility;
  severity  : severity;
  timestamp : Ptime.t;
  hostname  : string;
  tag       : string;
  content   : string
}

module Rfc3164_Timestamp = struct
  let int_of_month_name = function
    | "Jan" -> Some 1
    | "Feb" -> Some 2
    | "Mar" -> Some 3
    | "Apr" -> Some 4
    | "May" -> Some 5
    | "Jun" -> Some 6
    | "Jul" -> Some 7
    | "Aug" -> Some 8
    | "Sep" -> Some 9
    | "Oct" -> Some 10
    | "Nov" -> Some 11
    | "Dec" -> Some 12
    | _ -> None

  let month_name_of_int = function
    | 1  -> "Jan"
    | 2  -> "Feb"
    | 3  -> "Mar"
    | 4  -> "Apr"
    | 5  -> "May"
    | 6  -> "Jun"
    | 7  -> "Jul"
    | 8  -> "Aug"
    | 9  -> "Sep"
    | 10 -> "Oct"
    | 11 -> "Nov"
    | 12 -> "Dec"
    | _ -> failwith "Invalid month integer"

  let encode ts =
    let ((_, month, day), ((h, m, s), _)) = Ptime.to_date_time ts in
    Printf.sprintf "%s %.2i %.2i:%.2i:%.2i" (month_name_of_int month) day h m s

  let decode s year : (Ptime.t * string, [> `Msg of string ]) result =
    let tslen = 16 in
    match String.length s with
    | l when l < tslen ->
      Error (`Msg "timestamp too short, must be at least 16 bytes")
    | l ->
      let month = int_of_month_name @@ String.sub s 0 3 in
      let day = String.sub s 4 2 |> String.trim |> int_of_string_opt in
      let hour = String.sub s 7 2 |> int_of_string_opt in
      let minute = String.sub s 10 2 |> int_of_string_opt in
      let second = String.sub s 13 2 |> int_of_string_opt in
      match month, day, hour, minute, second with
      | None, _, _, _, _ -> Error (`Msg "couldn't decode month in timestamp")
      | _, None, _, _, _ -> Error (`Msg "couldn't decode day in timestamp")
      | _, _, None, _, _ -> Error (`Msg "couldn't decode hours in timestamp")
      | _, _, _, None, _ -> Error (`Msg "couldn't decode minutes in timestamp")
      | _, _, _, _, None -> Error (`Msg "couldn't decode seconds in timestamp")
      | Some month, Some day, Some hour, Some min, Some sec ->
        match Ptime.of_date_time ((year, month, day), ((hour, min, sec), 0)) with
        | None -> Error (`Msg "couldn't transform timestamp to ptime.t")
        | Some ts -> Ok (ts, String.sub s tslen (l - tslen))
end

let to_string msg =
  let facility = string_of_facility msg.facility
  and severity = string_of_severity msg.severity
  and timestamp = Rfc3164_Timestamp.encode msg.timestamp
  in
  Printf.sprintf
    "Facility: %s Severity: %s Timestamp: %s Hostname: %s Tag: %s Content: %s\n%!"
    facility severity timestamp msg.hostname msg.tag msg.content

let pp ppf msg = Format.pp_print_string ppf (to_string msg)

let encode_gen encode ?len msg =
  let facse = int_of_facility msg.facility * 8 + int_of_severity msg.severity
  and ts = Rfc3164_Timestamp.encode msg.timestamp
  in
  let msgstr = encode facse ts msg.hostname msg.tag msg.content
  in
  match len with
  | None -> msgstr
  | Some max_len ->
    if String.length msgstr > max_len then
      String.sub msgstr 0 max_len
    else
      msgstr

let is_alphanum = function
  | '0' .. '9' | 'A' .. 'Z' | 'a' .. 'z' -> true
  | _ -> false

let separator s =
  match String.get s 0 with
  | exception Invalid_argument _ -> " "
  | c when not (is_alphanum c) -> ""
  | _ -> " "

let encode ?len msg =
  let encode facse ts hostname tag content =
    Printf.sprintf "<%d>%s %s %s%s%s" facse ts hostname tag
      (separator content) content
  in
  encode_gen encode ?len msg

let encode_local ?len msg =
  let encode facse ts _ tag content =
    Printf.sprintf "<%d>%s %s%s%s" facse ts tag (separator content) content
  in
  encode_gen encode ?len msg

let parse_priority_value s :
  (facility * severity * string, [> `Msg of string ]) result =
  match String.split_on_char '<' s with
  | "" :: datas ->
    begin
      let data = String.concat "<" datas in
      match String.split_on_char '>' data with
      | pri :: datas ->
        begin
          let data = String.concat ">" datas in
          if String.length pri > 3 then
            Error (`Msg "couldn't parse priority: expected '>' earlier")
          else
            (* TODO RFC 3164 4.1.1 requires decimal, String.to_int accepts "0x1" *)
            match int_of_string_opt pri with
            | None -> Error (`Msg "couldn't parse priority: not an integer")
            | Some priority_value ->
              let facility = facility_of_int @@ priority_value / 8
              and severity = severity_of_int @@ priority_value mod 8
              in
              match facility, severity with
              | None, _ -> Error (`Msg "invalid facility")
              | _, None -> Error (`Msg "invalid severity")
              | Some facility, Some severity -> Ok (facility, severity, data)
        end
      | _ -> Error (`Msg "couldn't parse priority: no '>' found")
    end
  | _ -> Error (`Msg "couldn't parse priority: expected '<'")

let parse_hostname s (ctx : ctx) : (string * string, [> `Msg of string ]) result =
  if ctx.set_hostname then
    Ok (ctx.hostname, s)
  else
    match String.split_on_char ' ' s with
    | host :: datas ->
      let data = String.concat " " datas in
      let* hostname =
        match String.split_on_char ':' host with
        | [ "" ] -> Error (`Msg "invalid or empty hostname")
        | [ host ]
        | [ host ; "" ] -> Ok host
        | _ -> Error (`Msg "invalid empty hostname")
      in
      Ok (hostname, data)
    | _ -> Error (`Msg "invalid or empty hostname")

let parse_timestamp s (ctx : ctx) =
  let ((year, _, _), _) = Ptime.to_date_time ctx.timestamp in
  match Rfc3164_Timestamp.decode s year with
  | Ok (timestamp, data) -> Ok (timestamp, data, ctx)
  | Error _ ->
    let ctx = { ctx with set_hostname = true } in
    Ok (ctx.timestamp, s, ctx)

let parse_tag s : (string * string, [> `Msg of string ]) result =
  let rec collect s len idx =
    if idx > 32 then
      Result.Error (`Msg "tag exceeds 32 characters")
    else if idx >= len then
      Ok (s, "")
    else if is_alphanum (String.get s idx) then
      collect s len (idx + 1)
    else
      Ok (String.sub s 0 idx, String.sub s idx (len - idx))
  in
  collect s (String.length s) 0

(* FIXME Provide default Ptime.t? Version bellow doesn't work. Option type
let parse ?(ctx={timestamp=(Ptime.of_date_time ((1970, 1, 1), ((0, 0,0), 0))); hostname="-"; set_hostname=false}) data =
*)
let decode ~ctx data : (t, [> `Msg of string ]) result =
  let* facility, severity, data = parse_priority_value data in
  let* timestamp, data, ctx = parse_timestamp data ctx in
  let* hostname, data = parse_hostname data ctx in
  let* tag, content = parse_tag data in
  Ok { facility; severity; timestamp; hostname; tag; content }
